v2/cf-nodejs-logging-support/src/lib/logger/record-factory.ts

```diff
@@ -1,29 +1,58 @@
 import util from "util";
 import Config from "../config/config";
-import { ConfigField } from "../config/interfaces";
+import { Source } from "../config/interfaces";
 import NestedVarResolver from "../helper/nested-var-resolver";
 import RequestAccessor from "../middleware/request-Accessor";
 import ResponseAccessor from "../middleware/response-accessor";
 import ReqContext from "./context";
+import { SourceUtils } from "./source-utils";
+const stringifySafe = require('json-stringify-safe');
 
 export default class RecordFactory {
+    static MAX_STACKTRACE_SIZE = 55 * 1024;
     private static REDACTED_PLACEHOLDER = "redacted";
 
     // init a new record and assign fields with output "msg-log"
-    static buildMsgRecord(args: Array<any>, context?: ReqContext): any {
+    static buildMsgRecord(registeredCustomFields: Array<string>, loggerCustomFields: Map<string, any>, level: string, args: Array<any>, context?: ReqContext): any {
 
         const configInstance = Config.getInstance();
         const msgLogFields = configInstance.getMsgFields();
         let record: any = {
-            "level": "info",
+            "level": level,
         };
 
-        msgLogFields.forEach(field => {
-            if (field._meta!.isEnabled == false) {
-                return;
+        var customFieldsFromArgs = {};
+        var lastArg = args[args.length - 1];
+
+        if (typeof lastArg === "object") {
+            if (RecordFactory.isErrorWithStacktrace(lastArg)) {
+                record.stacktrace = RecordFactory.prepareStacktrace(lastArg.stack);
+            } else if (RecordFactory.isValidObject(lastArg)) {
+                if (RecordFactory.isErrorWithStacktrace(lastArg._error)) {
+                    record.stacktrace = RecordFactory.prepareStacktrace(lastArg._error.stack);
+                    delete lastArg._error;
+                }
+                customFieldsFromArgs = lastArg;
             }
+            args.pop();
+        }
 
-            record[field.name] = this.getFieldValue(field, record);
+        msgLogFields.forEach(field => {
+
+            if (!Array.isArray(field.source)) {
+                record[field.name] = this.getFieldValue(field.source, record);
+            } else {
+                let sourceIndex = 0;
+                while (!record[field.name]) {
+                    sourceIndex = SourceUtils.getNextValidSourceIndex(field.source, sourceIndex);
+                    if (sourceIndex == -1) {
+                        return;
+                    }
+                    let source = field.source[sourceIndex];
+                    record[field.name] = this.getFieldValue(source, record);
+                    ++sourceIndex;
+                }
+            }
```

Comment:
This section is duplicated. Can we move it to a common location and consume it?

Reply:
Fixed in https://github.com/SAP/cf-nodejs-logging-support/pull/121/commits/e265b467df8b9f140a55a728fb7a403909b9f12d.

Note that in total the new function `getValuesFromSources` is consumed in 3 different places:

1. When getting values for msg-log in `RecordFactory`
2. When getting values for req-log in `RecordFactory`
3. When getting values for context in `ReqContext`

Each case consider different sources, thats why there are 3 functions to get a fields value: 

1. `getFieldValue`
2. `getReqFieldValue`
3. `getContextField`

I grouped all these source related functions to get fields values in `SourceUtils`.